pkg/archive/zip.go

```diff
@@ -68,40 +70,156 @@ func (e ZipEntry) WriteTo(zw *zip.Writer) error {
-	defer zw.Close()
-	var ents []ZipEntry
-	for _, f := range zr.File {
-		r, err := f.Open()
+// MutableZipFile wraps zip.File to allow in-place modification of the original.
+type MutableZipFile struct {
+	zip.FileHeader
+	File       *zip.File
+	mutContent []byte
+}
+
+func (mf *MutableZipFile) Open() (io.Reader, error) {
+	if mf.mutContent != nil {
+		return bytes.NewReader(mf.mutContent), nil
+	}
+	return mf.File.Open()
+}
+
+func (mf *MutableZipFile) SetContent(content []byte) {
+	mf.mutContent = content
+}
+
+// MutableZipReader wraps zip.Reader to allow in-place modification of the original.
+type MutableZipReader struct {
+	*zip.Reader
+	File []*MutableZipFile
+}
+
+func NewMutableReader(zr *zip.Reader) MutableZipReader {
+	mr := MutableZipReader{Reader: zr}
+	for _, zf := range zr.File {
+		mr.File = append(mr.File, &MutableZipFile{File: zf, FileHeader: zf.FileHeader})
+	}
+	return mr
+}
+
+func (mr MutableZipReader) WriteTo(zw *zip.Writer) error {
+	for _, mf := range mr.File {
+		r, err := mf.Open()
 		if err != nil {
 			return err
 		}
-		b, err := io.ReadAll(r)
+		w, err := zw.CreateHeader(&mf.FileHeader)
 		if err != nil {
-			r.Close()
 			return err
 		}
-		if err := r.Close(); err != nil {
+		if _, err := io.Copy(w, r); err != nil {
 			return err
 		}
-		// TODO: Memory-intensive. We're buffering the full file in memory (again).
-		// One option would be to do two passes and only buffer what's necessary.
-		ents = append(ents, ZipEntry{&zip.FileHeader{Name: f.Name, Modified: time.UnixMilli(0)}, b})
 	}
-	sort.Slice(ents, func(i, j int) bool {
-		return ents[i].FileHeader.Name < ents[j].FileHeader.Name
-	})
-	for _, ent := range ents {
-		w, err := zw.CreateHeader(ent.FileHeader)
-		if err != nil {
-			return err
-		}
-		if _, err := io.Copy(w, bytes.NewReader(ent.Body)); err != nil {
-			return err
+	return nil
+}
+
+type ZipArchiveStabilizer struct {
+	Name string
+	Func func(*MutableZipReader)
+}
+
+type ZipEntryStabilizer struct {
+	Name string
+	Func func(*MutableZipFile)
+}
+
+var AllZipStabilizers []any = []any{
+	StableZipFileOrder,
+	StableZipModifiedTime,
+	StableZipCompression,
+	StableZipDataDescriptor,
+	StableZipFileEncoding,
+	StableZipFileMode,
+	StableZipMisc,
+}
+
+var StableZipFileOrder = ZipArchiveStabilizer{
+	Name: "zip-file-order",
+	Func: func(zr *MutableZipReader) {
+		slices.SortFunc(zr.File, func(i, j *MutableZipFile) int {
+			return strings.Compare(i.Name, j.Name)
+		})
+	},
+}
+
+var StableZipModifiedTime = ZipEntryStabilizer{
+	Name: "zip-modified-time",
+	Func: func(zf *MutableZipFile) {
+		zf.Modified = time.UnixMilli(0)
+		zf.ModifiedDate = 0
+		zf.ModifiedTime = 0
+	},
+}
+
+var StableZipCompression = ZipEntryStabilizer{
+	Name: "zip-compression",
+	Func: func(zf *MutableZipFile) {
+		zf.Method = zip.Store
+	},
+}
+
+var StableZipDataDescriptor = ZipEntryStabilizer{
+	Name: "zip-data-descriptor",
+	Func: func(zf *MutableZipFile) {
+		zf.Flags = zf.Flags & ^uint16(0x8)
```

Comment:
Could you add a comment for which flag this is changing?

Reply:
I mean, the third bit? the "data descriptor" flag? what do you mean?

Reply:
Yeah, I mean giving the name of the flag. The rest of the behavior is readable in English but zeroing the third bit seems like a magic number.

Reply:
eh actually naming it also enables us to do the cast once instead of twice. changed.